README.md
# FastORM
### ORM for async postgres
##### Beta

FastORM is a modern, fast (async), database library for projects with Python 3.10+ based on standard Python type hints.

The key features are:
 - Async postgres
 - Tested to work with Python 3.10

#### Install
```
pip install fastorm
```

#### Example
> See [example.py](example.py) for more examples.

Let's define some tables, to show off the capabilities.

```py
class State(str, Enum):
    RUNNING = 'running'
    ABORTED = 'stopped'
    COMPLETED = 'done'
# end class


class User(FastORM):
    _ignored_fields = []
    _primary_keys = ['id']
    _automatic_fields = ['id']
    _table_name = 'user'

    id: Optional[int]  # Optional because automatically filled (_automatic_fields)
    name: str


class Auction(FastORM):
    _ignored_fields = []
    _primary_keys = ['id']
    _automatic_fields = ['id']
    _table_name = 'auction'

    id: Optional[int]  # Optional because automatically filled (_automatic_fields)
    owner: Union[int, User]  # can be provided by int or the native object
    previous_owner: Optional[User]  # Optional because nullable
    state: State  # Enum support :D
    title: str
    subtitle: Optional[str]  # nullable
    description: str
    start_date: datetime  # datetime support
    end_date: datetime
    metadata: JSONType
    deleted: bool
    chat_id: int
```

Now you can quickly write classes to the database:

```py
conn = await FastORM.create_connection('postgresql://user:password@postgres_host/database')


user = User(id=None, name="hunter24")  # id will be filled by the database
await owner_user.insert(conn=conn)  # set's the id, too.

auction = Auction(
    id=None,  # gonna be automatic if `insert(…, ignore_setting_automatic_fields=False)` (default).
    # two ways of setting references to other tables:
    # by the actual value, in this case the numeric id
    owner=user.id,  
    # or via a different object,
    # it will use the id field (internally set by `User._primary_keys`) to determine the actual values.
    previous_owner=user,  
    state=State.RUNNING,  # enum will be a string in the database
    title="I sell my soul", subtitle="Slightly used",
    description="You only get a share though, others claim ownership, too.",
    start_date=datetime.now(), end_date=datetime.now() + timedelta(days=5),  # datetimes just works
    metadata={"just": ["json", "stuff", 111, datetime.now()]},  # will be native JSONB. You can have datetimes and your own classes in there as well, see `FastORM._set_up_connection`.
    deleted=False,
    chat_id=9223372036854775807,  # note, this database field must be BIGINT for such large numbers
)
await auction.insert(conn=conn)
```

Basic lookups are easy, too.

```py
# single lookup, returns one element or None
user = await User.get(name="hunter24")
user = await User.get(id=1234)

# list of results (list can have length 0)
all_running_auctions = Auction.select(state=State.RUNNING)
```

Of course updating and deleting is possible too.

```py
auction.state = State.COMPLETED
await auction.update()
```
```py
await user.delete()
```


CHANGELOG.md
> ### Known bugs:
> 1. In Python versions **before Python 3.10**, the automated **CREATE TABLE statement** (`build_sql_create(…)` thus also `create_table(…)`) will have problems with referencing other tables and list types:     
>   That means, `list[int]` will be missing an array at the end (`BIGINT[]` will be only `BIGINT`, `BIGINT[][][]` will be only `BIGINT[][]`) and
>   referencing other tables as part of an optional segment will result in an `JSONB` field instead of the referencing fields.   
>   **Workaround**: Change the type of the field after creation in python versions if you must use both 3.10 and the `create_table(…)` (or `build_sql_create(…)`) function
> - That's all we know of.


> ### Known limitations:
> 1. Currently references are not working correctly if you provide a union with it and the underlying primary type (or type tuple.)
>    **Solution**: Specify `foo: Bar` instead of `foo: Union[Bar, Tuple[int, int]]`.
> 2. Do not use `from __future__ import annotations`, as that turns all types to be strings, which currently can't be inspected on class creation. 

# Planned
- ~~Added sync version of the client~~

# Not yet released

# v0.0.16
- 🔨 Fixed a lot more stuff to make it now work quite nicely with FastAPI.
- 🆕 Added a new `consume_from_row(row)` to make custom query result parsing easier.
- 🔨 Fixed it not handling timezones in `datetime`s but instead breaking.
    - It now converts everything it encounters to UTC first.
    - For naïve `datetime` objects, that is assumed to take the computer's local timezone and is adapted for that to UTC as well. 
- 🆕 Take into account when you set `Field(default=123)` or `Field(default_factory=datetime.utcnow)` as value of the class field.
- 🔨 Fixed `In[value, None]`.
- 🔨 Fixed `field=None` in `select(…)` and `get(…)` not properly resulting in `field IS NULL`.

# v0.0.15
- 🔨 Some fixes making it work with FastAPI.

# v0.0.14
- 🔨 Fixed `REFERENCES` for more than one level of primary key references. 

# v0.0.13
- 🔄 Made sure that the `async insert(…)` method contains no database-agnostic code, and that is actually refactored out to calls to new `_insert_preparation` and `_insert_postprocess` methods.
    - This is in preparation of the sync version of the client.  
- 🆕 Make classes referencing themself possible.
- 🔨 Fixed `DELETE` not working for primary keys which are references.
- 🔨 Fixed `REFERENCES` to other tables with a double Primary key now set the `FOREIGN KEY` correctly in one, as opposed to trying to create `FOREIGN KEY` for each field.
    - The old generated SQL would result in `ERROR: there is no unique constraint matching given keys for referenced table "<table name>"`

# v0.0.12
- 🆕 Added a new `FastORM.get_primary_keys_sql_fields()` method to get the sql column name(s) of the primary key(s).
- ⚠️ Replaced the `on_conflict_upsert_field_list` parameter of `FastORM.insert(…)` and `FastORM.build_sql_insert(…)` with a new `upsert_on_conflict`.
   - It is mostly the same, but the variable name speaks better what it does. That is, specifying the fields which it would conflict on.
   - Migration:
     - `on_conflict_upsert_field_list=['some_field', 'another_field']` is now `upsert_on_conflict=['some_field', 'another_field']`.
     - `on_conflict_upsert_field_list=None` is now `upsert_on_conflict=False`.
   - 🆕 Also there's a new case now:
     - `upsert_on_conflict=True`. Here it will now automatically use the primary keys to conflict on those.
- 🔨 Fixed `FastORM.insert(…)` and `FastORM.build_sql_insert(…)` that a key listed in the conflict list for UPSERTs would not be excluded from the UPDATE part of the UPSERT, the value overwriting part.
    

# v0.0.11
- 🔨 Fixed equality checking on deduplication if only one of them was a `ForwardRef`.
- 🔨 Fixed upgrading of `Any` failing.


# v0.0.10
- 🔨 Fixed `SELECT` syntax for table references
- 🔨 Fixed `INSERT`, `UPDATE`, `SELECT`, `DELETE` for new field syntax (where a field referencing `OtherTable.id` would be called `other_table__id`).
- 🔨 Fixed `from_row(…)` if you used namespaces (those from `get_select_fields(namespace="…")`) 
- ✅ Improved test coverage to pretty much everything.


# v0.0.9
- 🆕 Added caching for `get_fields_typehints(…)` and `get_fields_references(…)`.
- 🔨 Fixed error if you used a `ForwardRef` to reference a model not yet defined.
- ⚠️ Extracted the referencing part from the create stable stuff (`create_table(…)` and `build_sql_create(…)`). That way they can be executed independently (`create_table_references(…)` and `build_sql_references(…)`), so you can first create all the tables and then add the references.

# v0.0.8
- 🆕 added `get_connection_pool(…)` to get a connection pool ready to use.
- ⚠️ renamed `get_connection(…)` to `create_connection(…)`. To fit the naming of `get_connection_pool(…)`.
- 🔨 Fixed invalid `SELECT` query if there were no parameters given to `.select(…)` (or the underlying `.build_sql_select(…)`).
    - Empty selectors will continue to be allowed for `.get(…)` as well, to keep that funny case possible, where there's only one row in a table and you wanna simply `SomeTable.get()` it. 

# v0.0.7
- 🔨 Fix for CREATE table statement not allowing for a prepared statement, and thus needing the escaping of values.
    - ℹ️ Needs psycopg2 installed for complex types (everything other than None, bool, int, and pure ascii strings)
- 🆕 Added support for `PRIMARY KEY`(`S`).
- 🆕 Added support for creating `REFERENCES` to other tables. 
     
# v0.0.6
- 🆕 Handle `Optional[Table, <Table Primary Key Type>]` for references.
    - Works for referencing other tables with a multi-field primary key, too. 
- Added lots of unit tests and also fix bugs found by this
    - 🆕 added support for multiple primary keys in most places now
    - 🔨 fixes for errors/wonkyness introduced by using pydantic's BaseModel.

# v0.0.5
- 🆕 Added `create_table(cls, conn: Connection, if_not_exists: bool = False)`.

example.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
__author__ = 'luckydonald'

import asyncio
from datetime import datetime, timedelta
from typing import Optional, Union
from enum import Enum

from asyncpg import Connection
from luckydonaldUtils.typing import JSONType

from fastorm import FastORM


class State(str, Enum):
    RUNNING = 'running'
    ABORTED = 'stopped'
    COMPLETED = 'done'
# end class


class User(FastORM):
    _ignored_fields = []
    _primary_keys = ['id']
    _automatic_fields = ['id']
    _table_name = 'user'

    id: Optional[int]  # Optional because automatically filled (_automatic_fields)
    name: str
# end class


class Auction(FastORM):
    _ignored_fields = []
    _primary_keys = ['id']
    _automatic_fields = ['id']
    _table_name = 'auction'

    id: Optional[int]  # Optional because automatically filled (_automatic_fields)
    owner: Union[int, User]  # can be provided by int or the native object
    previous_owner: Optional[User]  # Optional because nullable
    state: State  # Enum support :D
    title: str
    subtitle: Optional[str]  # nullable
    description: str
    start_date: datetime  # datetime support
    end_date: datetime
    metadata: JSONType
    deleted: bool
    chat_id: int
# end class


async def run():
    conn = await FastORM.create_connection('postgresql://user:password@postgres_host/database')

    await migrate_database(conn)  # a function (below) which prepares the database for use, creates tables and manages schema changes.

    owner_user = User(id=1234, name="The new auction owner.")
    await owner_user.insert(conn=conn, ignore_setting_automatic_fields=True)

    previous_owner = User(id=None, name="The previous auction owner.")
    await previous_owner.insert(conn=conn, ignore_setting_automatic_fields=False)

    print(f'Inserted owner user as id={owner_user.id}.')
    print(f'Inserted previous owner user as id={previous_owner.id}.')

    auction = Auction(
        id=None,  # gonna be automatic if `insert(…, ignore_setting_automatic_fields=False)` (default).
        owner=1234,  # by numeric id, aka. the actual field value
        previous_owner=previous_owner,  # use the id field (`User._primary_keys`) to determine the actual integer value.
        state=State.RUNNING,  # enum will be a string in the database
        title="I sell my soul", subtitle="Slightly used",
        description="You only get a share though.",
        start_date=datetime.now(), end_date=datetime.now() + timedelta(days=5),  # datetimes just works
        metadata={"just": ["json", "stuff", 111, datetime.now()]},  # will be native JSONB. You can have datetimes and your own classes in there as well, see `FastORM._set_up_connection`.
        deleted=False,
        chat_id=9223372036854775807,  # note, this database field must be BIGINT for such large numbers
    )

    await auction.insert(conn=conn, ignore_setting_automatic_fields=True, on_conflict_upsert_field_list=None,  write_back_automatic_fields=True)
    print(f'Inserted auction as id={auction.id}.')
    auction_id = auction.id

    # now let's retrieve some data from the database!

    db_auctions_by_user = await Auction.select(conn=conn, owner=1234)  # you can provide everything after the `conn=conn` to identify your entries. Returns a list, which can be empty.
    db_auction = await Auction.get(conn=conn, id=auction_id)  # same as select, but only allows a single result. None if not found. Raises an error if it's more than one match.
    print(db_auction.title)

    # let's change something.
    db_auction.state = State.COMPLETED
    await db_auction.update(conn=conn)

    # run custom queries

    # e.g. joining the users to that result and loading them in a single query
    # basically we use a prepared statement for variables.
    fetch_params = (
        f'SELECT '
        f' {Auction.get_select_fields(namespace="auction")}'  # first all fields from the auction
        f', {User.get_select_fields(namespace="owner")}'  # then all fields from the owner
        f' FROM'
        f'  {Auction.get_table()} AS "auction",'  # notice this is the same as the `namespace=…` above
        f'  {User.get_table()} AS "owner",'  # notice this is the same as the `namespace=…` above
        # now the join condition for those table
        f'WHERE "owner"."id" = "auction"."owner"'  # join user table with auction for owner
        # and more general condition we are interested right now
        f' AND "auction"."end_date" > CURRENT_TIMESTAMP'  # still running
        f' AND ("auction"."state" = $1 OR "auction"."id" = $2)'  # I don't know, but two parameters so I can demostrate that.
        f' ORDER BY "auction"."end_date"'  # ascending, so we have the soonest to end first
        ,
        State.COMPLETED,
        1234,
    )  # so to make it clear, syntax is ["SQL", $1, $2, …], list or tuple
    # now run that query
    fetch_results = await conn.fetch(*fetch_params)
    # now we can deconstruct that into rows
    auction_fields_len = Auction.get_select_fields_len()
    user_fields_len = User.get_select_fields_len()
    for row in fetch_results:
        # split the results into the proper lists for both elements
        auction_fields = row[0:auction_fields_len]
        user_fields = row[auction_fields_len:auction_fields_len + user_fields_len]

        # generate objects from it
        database_auction = Auction.from_row(auction_fields)
        database_user = User.from_row(user_fields)
        print(f'Auction {database_auction.title} from user {database_user.name}')
    # end for
# end def


async def migrate_database(conn: Connection):
    """
    Database migration, to set up all the tables we need. With versions so you can simply add a new version if you modify your models.
    Usually a separate migration file, just in the same one here for the sake of the tutorial being a single file.
    """
    await conn.execute(
        """
        CREATE TABLE IF NOT EXISTS "version" (
          "version" SERIAL PRIMARY KEY,
          "date" TIMESTAMP NOT NULL DEFAULT CURRENT_TIMESTAMP,
          "message" TEXT NOT NULL,
          "details" TEXT NOT NULL,
          "updated_from_version" INTEGER
        )
        """
    )
    # Select the version from the table and handle the table being just created as version 0.
    now = datetime.now()
    row = await conn.fetchrow('SELECT "version", "date", "message", "details" FROM "version" ORDER BY "version" DESC')
    if row is None:
        version = None
        print(f'database no version yet, assuming version 0.')
    else:
        version, date, message, details = row
        print(f'database has version {version} from {date}: {message} ({details})')
    # end if
    updated_from_version = version

    if version < 1:
        version = 1
        message = "Create a first update, with a user table"
        details = "Just an user for now and the auction table later for the purpose of a tutorial"
        await conn.execute(
            """
            CREATE TABLE "user" (
                "id" SERIAL PRIMARY KEY,
                "name" TEXT NOT NULL
            )
            """
        )
        await conn.execute(
            '''INSERT INTO version(version, date, message, details, updated_from_version) VALUES($1, $2, $3, $4, $5)''',
            version, now, message, details, updated_from_version
        )
    # end if

    if version < 2:
        version = 2
        message = "Create a second update, with the auction table"
        details = None
        await conn.execute(
            """
            CREATE TABLE IF NOT EXISTS "auction" (

              "id" SERIAL PRIMARY KEY,  -- id: Optional[int]  # Optional because automaticly filled (_automatic_fields)
              "category" TEXT NOT NULL,
              "owner" INTEGER NOT NULL,  -- owner: Union[int, User]  # can be provided by int or the native object
              "previous_owner" INTEGER,  -- previous_owner: Optional[User]  # Optional because nullable
              "state" TEXT NOT NULL, -- state: State  # Enum support :D
              "title" TEXT NOT NULL, -- title: str
              "subtitle" TEXT, -- subtitle: Optional[str]  # nullable
              "description" TEXT NOT NULL, -- description: str
              "start_date" TIMESTAMP NOT NULL, -- start_date: datetime  # datetime support
              "end_date" TIMESTAMP NOT NULL, -- end_date: datetime
              "metadata" JSONB NOT NULL, -- metadata: JSONType
              "deleted" BOOLEAN NOT NULL, -- deleted: bool
              "chat_id" BIGINT NOT NULL  -- chat_id: int
            );
            """
        )
        await conn.execute(
            '''INSERT INTO version(version, date, message, details, updated_from_version) VALUES($1, $2, $3, $4, $5)''',
            version, now, message, details, updated_from_version
        )
    # end if
    print(f'Updated database from {updated_from_version} to {version}')
# end def


# Spin up a quick and simple event loop
# and run until completed
loop = asyncio.get_event_loop()
try:
    loop.run_until_complete(run())
finally:
    loop.close()
# end try
